tests/test_statemachine_bounded_transitions.py

```diff
@@ -0,0 +1,83 @@
+# coding: utf-8
+
+import mock
+import pytest
+
+from statemachine import StateMachine, State
+
+
+class MyModel(object):
+    def __init__(self, **kwargs):
+        for k, v in kwargs.items():
+            setattr(self, k, v)
+
+    def __repr__(self):
+        return "{}({!r})".format(type(self).__name__, self.__dict__)
+
+
+class CampaignMachine(StateMachine):
+    draft = State("Draft", initial=True)
+    producing = State("Being produced")
+    closed = State("Closed")
+
+    add_job = draft.to(draft) | producing.to(producing)
+    produce = draft.to(producing)
+    deliver = producing.to(closed)
+
+    def on_enter_producing(self, **kwargs):
+        pass
+
+    def on_exit_draft(self, **kwargs):
+        pass
+
+    def on_enter_closed(self):
+        pass
+
+    def on_exit_producing(self):
+        pass
+
+
+@pytest.fixture()
+def on_enter_mock():
+    with mock.patch.object(CampaignMachine, "on_enter_producing") as m:
+        yield m
+
+
+@pytest.fixture()
+def on_exit_mock():
+    with mock.patch.object(CampaignMachine, "on_exit_draft") as m:
+        yield m
+
+
+@pytest.fixture()
+def on_enter_backwards_mock():
+    with mock.patch.object(CampaignMachine, "on_enter_closed") as m:
+        yield m
+
+
+@pytest.fixture()
+def on_exit_backwards_mock():
+    with mock.patch.object(CampaignMachine, "on_exit_producing") as m:
+        yield m
+
+
+def test_run_transition_pass_arguments_to_sub_transitions(
+    on_enter_mock,
+    on_exit_mock,
+    on_enter_backwards_mock,
+    on_exit_backwards_mock,
+):
+    model = MyModel(state="draft")
+    machine = CampaignMachine(model)
+
+    machine.run("produce", param1="value1", param2="value2")
+
+    assert model.state == "producing"
+    on_enter_mock.assert_called_with(param1="value1", param2="value2")
+    on_exit_mock.assert_called_with(param1="value1", param2="value2")
+
+    machine.run("deliver", param3="value3")
+
+    # backwards compatibility
+    on_enter_backwards_mock.assert_called_with(param3="value3")
```

Comment:
Issue: I think that this way the actual method will not be called, so what I'm expecting to assert is that if you call the `run` with arguments, the handlers of events `on_enter_*` and `on_exit_*` that don't expect arguments will be fine. 

Since you've mocked the handlers, we cannot take this for true. 

**suggestion:** Consider defining the state machine under a fixture, and using a mock internally.

Reply:
got it! I have an idea

Reply:
I think that something like this will work:

``` python
@pytest.fixture
def state_machine():
    event_mock = mock.MagicMock()

    class CampaignMachine(StateMachine):
        draft = State("Draft", initial=True)
        producing = State("Being produced")
        closed = State("Closed")

        event_mock = event_mock

        add_job = draft.to(draft) | producing.to(producing)
        produce = draft.to(producing)
        deliver = producing.to(closed)

        def on_enter_producing(self, **kwargs):
            event_mock.on_enter_producing(**kwargs)

        def on_exit_draft(self, **kwargs):
            event_mock.on_exit_draft(**kwargs)

        def on_enter_closed(self):
            event_mock.on_enter_closed(*)

        def on_exit_producing(self):
            event_mock.on_exit_producing()

    return CampaignMachine
```